streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -408,26 +390,40 @@ trait ZSink[-R, +E, +A0, -A, +B] { self =>
     filter(a => !f(a))
 
   /**
-   * Effectfully ilters this sink by the specified predicate, dropping all elements for
+   * Effectfully filters this sink by the specified predicate, dropping all elements for
    * which the predicate evaluates to true.
    */
   final def filterNotM[E1 >: E, A1 <: A](f: A1 => IO[E1, Boolean]): ZSink[R, E1, A0, A1, B] =
     filterM(a => f(a).map(!_))
 
-  final def keyed[A1 <: A, K](f: A1 => K): ZSink[R, E, (K, A0), A1, Map[K, B]] =
-    new ZSink[R, E, (K, A0), A1, Map[K, B]] {
-      type State = Map[K, self.State]
+  /**
+   * Runs `n` sinks in parallel, where `n` is the number of possible keys
+   * generated by `f`.
+   */
+  final def keyed[A1 <: A, K](f: A1 => K): ZSink[R, E, (K, Chunk[A0]), A1, Map[K, B]] =
+    new ZSink[R, E, (K, Chunk[A0]), A1, Map[K, B]] {
+      type State = (Map[K, self.State], self.State)
 
-      val initial: ZIO[R, E, Step[State, Nothing]] =
-        self.initial.map(Step.leftMap(_)(default => Map[K, self.State]().withDefaultValue(default)))
+      val initial =
+        self.initial.map(init => (Map.empty[K, self.State].withDefaultValue(init), init))
 
-      def step(state: State, a: A1): ZIO[R, E, Step[State, (K, A0)]] = {
+      def step(state: State, a: A1) = {
         val k = f(a)
-        self.step(state(k), a).map(Step.bimap(_)(s1 => state + (k -> s1), b => (k, b)))
+        self.step(state._1(k), a).map(s => (state._1 + (k -> s), s))
       }
 
-      def extract(state: State): ZIO[R, E, Map[K, B]] =
-        ZIO.foreach(state.toList)(s => self.extract(s._2).map((s._1 -> _))).map(_.toMap)
+      def extract(state: State) =
+        ZIO
+          .foreach(state._1.toList) {
+            case (k, s) => self.extract(s).map(k -> _)
+          }
+          .map { list =>
+            val results   = list.map { case (k, (b, _))                        => (k, b) }.toMap
+            val leftovers = Chunk.fromIterable(list.map { case (k, (_, chunk)) => (k, chunk) })
+            (results, leftovers)
+          }
+
+      def cont(state: State) = self.cont(state._2)
```

Comment:
Interesting. The decision being made by the last state to step is a bit arbitrary.

I think this should be `state._2.values.exists(self.cont)` or `state._2.values.forall(self.cont)`. I lean towards `forall`.

Reply:
Then you can get rid of the right-hand side of the state tuple.

Reply:
Done.